Add jwt-parser function v1.0.0 for JWT token parsing in SonataFlow

ishanjogi89 opened the pull request:
# Pull Request Template

Thank you for your contribution to the Serverless Workflow Catalog! To help us review and merge your pull request, please provide the following information:

## Description

**Briefly describe the changes made in this pull request:**

This PR adds a new JWT token parsing function to the SonataFlow functions catalog. The function enables workflows to parse JWT tokens and extract user claims for personalized workflow responses, supporting the implementation of apache/incubator-kie-issues#1899.

## Related Issues

**Link to related issues or tasks:**

- apache/incubator-kie-issues#1899
- Related implementation PR: [Link to incubator-kie-kogito-runtimes PR]

## Type of Change

**Please check the type of change your pull request introduces:**

- [ ] Bug fix
- [x] New function
- [ ] Enhancement to existing function
- [ ] Documentation update
- [ ] Other (please describe)

## Function/Version Affected

**Specify the function and version affected by this pull request (if applicable):**

- **Function Name:** jwt-parser
- **Function Version:** 1.0.0

## Implementation Details

**Provide a summary of the implementation details:**

- **New Features:** 
  - JWT token parsing function with three operations: `parse`, `extractUser`, `extractClaim`
  - Support for extracting complete JWT payload as JSON
  - Support for extracting standard user claims (sub, preferred_username, email, etc.)
  - Support for extracting specific claims by name
  - Automatic handling of "Bearer " token prefix
  - Integration with SonataFlow workflow headers via `$WORKFLOW.headers`

- **Bug Fixes:** N/A

- **Changes to Existing Functions:** N/A

- **Dependencies Added:** 
  - Function calls the `jwt-parser` custom function type in SonataFlow runtime
  - No external dependencies for the catalog entry itself

## Additional Information

**Provide any additional information or context related to this pull request:**

This function specification supports the exact use case described in apache/incubator-kie-issues#1899, where workflows need to extract user information from JWT tokens for personalized responses. 

**Example usage from the issue:**
```json
{
  "functions": [
    {
      "name": "extractUser",
      "type": "custom",
      "operation": "jwt-parser:extractUser"
    }
  ],
  "states": [
    {
      "name": "extractUserName",
      "actions": [
        {
          "functionRef": {
            "refName": "extractUser",
            "arguments": {
              "token": "${ $WORKFLOW.headers.\"X-Authorization-acme_financial_auth\" }"
            }
          }
        }
      ]
    }
  ]
}
```

The function is implemented as a Quarkus addon in the SonataFlow runtime and this catalog entry provides the standard specification for its usage.

## Checklist

**Please ensure your pull request meets the following criteria:**

- [x] Code follows the project's coding style and guidelines
- [x] Changes are thoroughly tested
- [x] Documentation is updated (if applicable)

---

Thank you for contributing to the Serverless Workflow Catalog! Your efforts are greatly appreciated.

ishanjogi89 committed:
introducing jwt-parser function

Signed-off-by: Ishan Jogi <ijogi@redhat.com>

cdavernas commented:
@ishanjogi89 The function catalog applies to Serverless Workflow 1.x.x and above, but it looks like you’re mixing versions in your function. Also, the function should be calling a task internally, whereas yours is calling itself—which obviously won’t work.

To fix this, your function should actually be a set task that assigns a claims property using a jq expression to extract the claims from the token provided as input.

ishanjogi89 committed:
1. **Use Serverless Workflow 1.x syntax** with proper YAML format and `dsl: 1.0.0-alpha1`
2. **Replace the recursive call** with a proper `set` task using jq expressions
3. **Implement JWT decoding with jq**: The function now uses `split(".")[1] | @base64d | fromjson` to decode the JWT payload
4. **Add optional claim extraction** via `claimPath` parameter using jq path navigation

Signed-off-by: Ishan Jogi <ijogi@redhat.com>

ishanjogi89 committed:
updated Readme.md file

Signed-off-by: Ishan Jogi <ijogi@redhat.com>

ishanjogi89 commented:
Thank you for the feedback, @cdavernas ! You're absolutely right about the version mismatch and recursive call issue. I've updated the function specification to:

1. **Use Serverless Workflow 1.x syntax** with proper YAML format and `dsl: 1.0.0-alpha1`
2. **Replace the recursive call** with a proper `set` task using jq expressions
3. **Implement JWT decoding with jq**: The function now uses `split(".")[1] | @base64d | fromjson` to decode the JWT payload
4. **Add optional claim extraction** via `claimPath` parameter using jq path navigation

The function now properly decodes JWT tokens using pure jq expressions and can extract specific claims as requested.

ishanjogi89 committed:
covered edge case scenarios

Signed-off-by: Ishan Jogi <ijogi@redhat.com>

ishanjogi89 committed:
updated owners
Signed-off-by: Ishan Jogi <ijogi@redhat.com>

Signed-off-by: Ishan Jogi <ijogi@redhat.com>

cdavernas commented on functions/jwt-parser/1.0.0/README.md:
```diff
@@ -0,0 +1,192 @@
+  version: 1.0.0
+do:
+  - decodeJWT:
+      use: jwt-parser
```
That should actually be `call` 😉

cdavernas commented on functions/jwt-parser/1.0.0/README.md:
```diff
@@ -0,0 +1,192 @@
+  version: 1.0.0
+do:
+  - extractSubject:
+      use: jwt-parser
```
Should be `call`, too.

cdavernas commented on functions/jwt-parser/1.0.0/README.md:
```diff
@@ -0,0 +1,192 @@
+        claimPath: ".sub"
+        # Returns: { claims: {...}, result: "user-id-123" }
+  - extractNestedClaim:
+      use: jwt-parser
```
Should be `call`, too.

cdavernas commented on functions/jwt-parser/1.0.0/README.md:
```diff
@@ -0,0 +1,192 @@
+  version: 1.0.0
+do:
+  - parseRawToken:
+      use: jwt-parser
```
Should be `call`, too.

cdavernas commented on functions/jwt-parser/1.0.0/README.md:
```diff
@@ -0,0 +1,192 @@
+      with:
+        token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0.signature"
+  - parseBearerToken:
+      use: jwt-parser
```
Should be `call`, too.

cdavernas commented on functions/jwt-parser/1.0.0/function.yaml:
```diff
@@ -0,0 +1,39 @@
+    document:
+      type: object
+      description: The parsed JWT payload or extracted claim as a JSON object
+use: set
```
`use` is not part of SW 1.0.0, in that context. You can remove it 😉

cdavernas reviewed (changes requested):
There's still some minor fixes to be made, but thank you for your awesome work! ❤️

cdavernas commented:
@ishanjogi89 Don't you think it would be better if the function returned either an array containing the extracted claims, when `claimPath` is set, or the value of the retrieved claim, when `claimPath` has been set? It would be, IMHO, a bit leaner and more intuitive, but that's just my opinion 😉

cdavernas commented:
@JBBianchi @ricardozanini would you be so kind to review the proposal?

ricardozanini reviewed (commented):
@ishanjogi89 Thanks for the contribution. This is interesting.

I think a JWT parser can find some useful use cases to many implementations out there.

You mentioned SonataFlow in your PR description. I'm a maintainer of this project, and we are not currently working on upgrading the runtime engine to SWF 1.0.0. 

I don't think this function will help you achieve what you need on the implementation side. Unfortunately, version 0.8 has many limitations - one of the reasons why we upgraded the spec.

However, you can add an extension as a custom function to the SonataFlow engine here: https://github.com/apache/incubator-kie-kogito-runtimes/tree/main/kogito-serverless-workflow.

Regarding the PR, once @cdavernas' comments are handled, it should be fine. Great addition, thanks again!

ishanjogi89 committed:
fixed issue as per review comments

Signed-off-by: Ishan Jogi <ijogi@redhat.com>
